src/game/logic/object/weaponset.cpp

```diff
@@ -0,0 +1,259 @@
+    "SUBDUAL_BUILDING",
+    "SUBDUAL_UNRESISTABLE",
+    "MICROWAVE",
+    "KILL_GARRISONED",
+    "STATUS",
+    nullptr,
+};
+
+WeaponSet::WeaponSet() :
+    m_curWeapon(WEAPONSLOT_PRIMARY),
+    m_curWeaponLocked(0),
+    m_curWeaponTemplateSet(nullptr),
+    m_filledWeaponSlotMask(0),
+    m_pitchLimited(false),
+    m_damageWeapon(false)
+{
+    m_damageTypes.Clear();
+
+    for (int i = 0; i < WEAPONSLOT_COUNT; i++) {
+        m_weapons[i] = nullptr;
+    }
+}
+
+WeaponSet::~WeaponSet()
+{
+    for (int i = 0; i < WEAPONSLOT_COUNT; i++) {
+        if (m_weapons[i] != nullptr) {
+            m_weapons[i]->Delete_Instance();
+        }
+    }
+}
+
+void WeaponSet::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+
+    if (xfer->Get_Mode() == XFER_LOAD) {
+        Utf8String name;
+        BitFlags<WEAPONSET_COUNT> flags;
+        xfer->xferAsciiString(&name);
+        flags.Xfer(xfer);
+
+        if (name.Is_Empty()) {
+            m_curWeaponTemplateSet = nullptr;
+        } else {
+            ThingTemplate *tmplate = g_theThingFactory->Find_Template(name, true);
+
+            if (tmplate == nullptr) {
+                throw CODE_06;
+            }
+
+            m_curWeaponTemplateSet = tmplate->Find_Weapon_Template_Set(flags);
+
+            if (m_curWeaponTemplateSet == nullptr) {
+                throw CODE_06;
+            }
+        }
+    } else if (xfer->Get_Mode() == XFER_SAVE) {
+        Utf8String name;
+        BitFlags<WEAPONSET_COUNT> flags;
+
+        if (m_curWeaponTemplateSet != nullptr) {
+            const ThingTemplate *tmplate = m_curWeaponTemplateSet->Get_Type();
+
+            if (tmplate == nullptr) {
+                throw CODE_06;
+            }
+
+            name = tmplate->Get_Name();
+            flags = m_curWeaponTemplateSet->Get_Conditions();
+        }
+
+        xfer->xferAsciiString(&name);
+        flags.Xfer(xfer);
+    }
+
+    for (int i = 0; i < WEAPONSLOT_COUNT; i++) {
+        bool has_weapon = m_weapons[i] != nullptr;
+        xfer->xferBool(&has_weapon);
+
+        if (has_weapon) {
+            if (xfer->Get_Mode() == XFER_LOAD && !m_weapons[i]) {
+                WeaponTemplate *weapon = m_curWeaponTemplateSet->Get_Nth(static_cast<WeaponSlotType>(i));
+
+                if (weapon == nullptr) {
+                    captainslog_dbgassert(false, "xfer backwards compatibility code - old save file???");
+                    weapon = m_curWeaponTemplateSet->Get_Nth(WEAPONSLOT_PRIMARY);
+                }
+
+                m_weapons[i] = WeaponStore::Allocate_New_Weapon(weapon, static_cast<WeaponSlotType>(i));
+            }
+
+            m_weapons[i]->Xfer_Snapshot(xfer);
+        }
+    }
+
+    xfer->xferUser(&m_curWeapon, sizeof(m_curWeapon));
+    xfer->xferUser(&m_curWeaponLocked, sizeof(m_curWeaponLocked));
+    xfer->xferUnsignedInt(&m_filledWeaponSlotMask);
+    xfer->xferInt(&m_totalAntiMask);
+    xfer->xferBool(&m_damageWeapon);
```

Comment:
m_damageWeapon is serialized twice. Is this meant to be m_pitchLimited?

Reply:
This is legit and what the original code does.